### DISCUSSION TITLE
Anyone here see this before? - Configuring MDM push services

### DISCUSSION THREAD

**Discussion by motionbug** [cirruslabs/tart#958] [General]
**Context:** 
**Building Custom Packer Image with Tart Provisioner**

Following the completion and upload of the build to the OCI repository, the image is downloaded, cloned, and then configured with Tart to create a new Mac and serial number for the virtual machine.

```tart set --random-mac --random-serial <nameofvm>```

The VM is enrolled, but an issue arises. Although the MDM profile is installed, it is stalled. Upon a reboot, the following startup message is displayed: after this message, the MDM profile is functional, but the device is incompletely enrolled.

This behaviour does not occur when using VM images from the Tart public registry.

Could anyone provide an explanation or have encountered this issue before?

![image](https://github.com/user-attachments/assets/10d2c769-f92e-41c3-9473-6f2d7e5a58b4)

**Comment by edigaryev** [comment]
Hi Rob,

Would you mind showing some more steps to reproduce this? I.e. how do you enroll a VM and install an MDM profile?

**Reply by motionbug** [reply]
Thanks @edigaryev 

The ghcr.io/cirruslabs/macos-sequoia-vanilla:latest image have a rotating UUID, on the packer builds that I make, even with 

--random-mac --randon-serial

the UUID of the device stays the same. 

What am I missing, is there some other thing that you are doing to build the packer images where the UUIDs are removed after build?

Thanks

**Reply by edigaryev** [reply]
>The ghcr.io/cirruslabs/macos-sequoia-vanilla:latest image have a rotating UUID

I've tried the following, making sure that no other local VMs are present on the host before running this:

```shell
tart pull ghcr.io/cirruslabs/macos-sequoia-vanilla:latest
tart clone ghcr.io/cirruslabs/macos-sequoia-vanilla:latest macos1
tart clone ghcr.io/cirruslabs/macos-sequoia-vanilla:latest macos2
tart run macos1 &
tart run macos2 &
```

(`ghcr.io/cirruslabs/macos-sequoia-vanilla:latest` resolved to `ghcr.io/cirruslabs/macos-sequoia-vanilla@sha256:55ec8579335884de600c515ae210b9c264cbb0949c771a88ba62264735dcfc9d` on `tart pull`)

The two VMs started and had different MAC-addresses, but identical serial number, hardware UUID and a provisioning UUID.

Now, if I re-try with `--random-serial`:

```
tart delete macos1 macos2
tart clone ghcr.io/cirruslabs/macos-sequoia-vanilla:latest macos1
tart clone ghcr.io/cirruslabs/macos-sequoia-vanilla:latest macos2
tart set macos1 --random-mac --random-serial
tart set macos2 --random-mac --random-serial
tart run macos1 &
tart run macos2 &
```

All three identifiers are different, just as expected.

**Reply by motionbug** [reply]
Haha thank you @edigaryev I was about to post something exactly like this.

I have built my own packer template to "prep" and image so when pulled it has things in place for the test enrolment.
When I push an image to my OCI server and then pull it down, the identifiers are same

How do you build yours so they can be different?

**Reply by edigaryev** [reply]
>How do you build yours so they can be different?

https://github.com/cirruslabs/macos-image-templates

**Reply by robatjamf** [reply]
These are the templates that I am using! And yet I have the same hardware ID while the ones I pull from ghcr.io/cirruslabs/macos-sequoia-vanilla:latest, when cloned do not have the same hardware ID

**Reply by edigaryev** [reply]
OK, would you mind going through a checklist in https://github.com/cirruslabs/tart/discussions/958#discussioncomment-11447179 and reporting your findings?

**Comment by motionbug** [comment]
So when I pull the tart image from **ghcr.io/cirruslabs/macos-sequoia-vanilla:latest** 

Looking at the **config.json** at the root of ghcr.io/cirruslabs/macos-sequoia-vanilla:latest the hardware ID: _"YnBsaXN0MDDTAQIDBA-blabla-AAAAAAAAAkAAAAAAAAAAAAAAAAAAABt"_

I clone from ghcr.io/cirruslabs/macos-sequoia-vanilla:latest to `macos2`

Looking at the **config.json** at the root of macos2  the hardware ID is different!! 
_"YnBsaXN0MDDTAQIDBAUGJ-blabla-ZaXAAAAAAAAAEBAAAAAAAAAAkAAAAAAAAAAAAAAAAAAABe"_

I clone my packer image the hardware id of my OCI and my clone are the same. That is the error I am getting and I am wondering what makes your image different than mine @edigaryev 

Any insight to this issue would help me out! What you do vs what I am doing!

**Reply by edigaryev** [reply]
>Looking at the config.json at the root of ghcr.io/cirruslabs/macos-sequoia-vanilla:latest the hardware ID: "YnBsaXN0MDDTAQIDBA-blabla-AAAAAAAAAkAAAAAAAAAAAAAAAAAAABt"

1. Are you looking at the `ecid` field in the `config.json` blob in the OCI registry, or on a local machine?
    1. If that's OCI registry, which command or method are you using to obtain this hardware ID you've posted? I'm seeing a different ECID at the time of writing this message:

        ```
        % brew install crane
        [...]
        % crane manifest ghcr.io/cirruslabs/macos-sequoia-vanilla:latest | jq -r '.layers[] | select(.mediaType == "application/vnd.cirruslabs.tart.config.v1") | .digest'
        sha256:6a0da6335c7a0a5cd335053c45f49a74fd4b863b04277265f1f17574d1c978f1
        % crane blob ghcr.io/cirruslabs/macos-sequoia-vanilla:latest@sha256:6a0da6335c7a0a5cd335053c45f49a74fd4b863b04277265f1f17574d1c978f1 | jq -r '.ecid'
        YnBsaXN0MDDRAQJURUNJRBMBZQfzl3IEEAgLEAAAAAAAAAEBAAAAAAAAAAMAAAAAAAAAAAAAAAAAAAAZ
        ```
    1. If that's local machine, what's the output of `tart fqn ghcr.io/cirruslabs/macos-sequoia-vanilla:latest` and `` commands? I'm seeing this at the time of writing this message, which also yields a different ECID that the one you're reporting:

        ```
        % tart fqn ghcr.io/cirruslabs/macos-sequoia-base:latest
        ghcr.io/cirruslabs/macos-sequoia-base@sha256:b92be3daa1d7d92c90f5674d9b24fbcdee96f10c297dfed84e921041738622cf
        % cat ~/.tart/cache/OCIs/ghcr.io/cirruslabs/macos-sequoia-vanilla/latest/config.json | jq -r '.ecid'
        YnBsaXN0MDDRAQJURUNJRBMBZQfzl3IEEAgLEAAAAAAAAAEBAAAAAAAAAAMAAAAAAAAAAAAAAAAAAAAZ
        ```

1. Which Tart version are you using (`tart --version`)?

1. How do you "clone from ghcr.io/cirruslabs/macos-sequoia-vanilla:latest to `macos2`", exactly? Can you try running the following commands, what would be the output:

    ```
    cat ~/.tart/cache/OCIs/ghcr.io/cirruslabs/macos-sequoia-vanilla/latest/config.json | jq -r '.ecid'
    tart clone ghcr.io/cirruslabs/macos-sequoia-vanilla:latest macos
    cat ~/.tart/vms/macos/config.json | jq -r '.ecid'
    ```

**Reply by robatjamf** [reply]
thanks @edigaryev we are confusing the ECID and the id that is associated with the hardwareModel

when enrolling into a MDM server we need a unique UUID and the hardwareModel seems to be the key here.

Steps I do to clone are:

---    
I do a `tart pull ghcr.io/cirruslabs/macos-sequoia-vanilla:latest` macOStest

then I do a `tart set macOStest --random-mac --random-serial`

the hardwareModel will be different for both VMs... and I can only get that to work with your images and I am using your packer templates but any clone I make from my finished packer build will have the same hardwareModel and will not enroll.

Really I have no idea what makes your images different when I do a clone and get a new hardwareModel

**Reply by robatjamf** [reply]
my fault too calling it 'hardware id'

**Reply by edigaryev** [reply]
>the hardwareModel will be different for both VMs... 

Are you talking about `hardwareModel` field in VMs directory's `config.json`?

If so, then I'm not seeing this behavior on my machine:

```shell
% tart pull ghcr.io/cirruslabs/macos-sequoia-vanilla:latest
pulling ghcr.io/cirruslabs/macos-sequoia-vanilla:latest...
pulling manifest...
ghcr.io/cirruslabs/macos-sequoia-vanilla@sha256:55ec8579335884de600c515ae210b9c264cbb0949c771a88ba62264735dcfc9d image is already cached and linked!
% cat ~/.tart/cache/OCIs/ghcr.io/cirruslabs/macos-sequoia-vanilla/latest/config.json | jq -r '.hardwareModel' | base64 -d | plutil -convert json - -o - | jq
{
  "DataRepresentationVersion": 1,
  "PlatformVersion": 2,
  "MinimumSupportedOS": [
    13,
    0,
    0
  ]
}
% tart clone ghcr.io/cirruslabs/macos-sequoia-vanilla:latest macos
% tart set --random-mac --random-serial macos
% cat ~/.tart/vms/macos/config.json | jq -r '.hardwareModel' | base64 -d | plutil -convert json - -o - | jq                                                 
{
  "DataRepresentationVersion": 1,
  "PlatformVersion": 2,
  "MinimumSupportedOS": [
    13,
    0,
    0
  ]
}
```

Could you reproduce these exact steps on yours and post results?

**Comment by motionbug** [comment]
Yes! Deleted everything that I had locally so a `tart list` command comes back with nothing.

**Step 1)**

```
tart pull ghcr.io/cirruslabs/macos-sequoia-vanilla:latest
pulling ghcr.io/cirruslabs/macos-sequoia-vanilla:latest...
pulling manifest...
```

**Step 2)**

```
tart clone ghcr.io/cirruslabs/macos-sequoia-vanilla:latest mynewtestmac
```

**Step 3)**

```
cat ~/.tart/cache/OCIs/ghcr.io/cirruslabs/macos-sequoia-vanilla/latest/config.json | jq -r '.hardwareModel'
YnBsaXN0MDDTAQIDBAUGXxAZRGF0YVJlcHJlc2VudGF0aW9uVmVyc2lvbl8QD1BsYXRmb3JtVmVyc2lvbl8QEk1pbmltdW1TdXBwb3J0ZWRPUxQAAAAAAAAAAAAAAAAAAAABEAKjBwgIEA0QAAgPKz1SY2VpawAAAAAAAAEBAAAAAAAAAAkAAAAAAAAAAAAAAAAAAABt

cat ~/.tart/vms/mynewtestmac/config.json | jq -r '.hardwareModel'
YnBsaXN0MDDTAQIDBAUGXxAZRGF0YVJlcHJlc2VudGF0aW9uVmVyc2lvbl8QD1BsYXRmb3JtVmVyc2lvbl8QEk1pbmltdW1TdXBwb3J0ZWRPUxQAAAAAAAAAAAAAAAAAAAABEAKjBwgIEA0QAAgPKz1SY2VpawAAAAAAAAEBAAAAAAAAAAkAAAAAAAAAAAAAAAAAAABt
```

Step 4)

```
tart set mynewtestmac --random-serial --random-mac
cat ~/.tart/vms/mynewtestmac/config.json | jq -r '.hardwareModel'
YnBsaXN0MDDTAQIDBAUGXxAZRGF0YVJlcHJlc2VudGF0aW9uVmVyc2lvbl8QD1BsYXRmb3JtVmVyc2lvbl8QEk1pbmltdW1TdXBwb3J0ZWRPUxABEAKjBwgIEA0QAAgPKz1SVFZaXAAAAAAAAAEBAAAAAAAAAAkAAAAAAAAAAAAAAAAAAABe
```

That is the steps I take and the images that I get form you are the only ones that change the `hardwareModel`. Any of my images that hardwareModel stays the same.

**Reply by motionbug** [reply]
how do we create a new `hardwareModel` then? I don't know enough as you do but I can only get them enrolled when its different.

**Reply by edigaryev** [reply]
OK, now I see.

The new `hardwareModel` gets created because we re-instantiate a [`VZMacHardwareModel`](https://developer.apple.com/documentation/virtualization/vzmachardwaremodel) on `tart set --random-serial`, and it seems that Apple had changed its data representation slightly, so now they're serializing this value differently.

It could be stuffing some host-specific data and whatnot, we don't know.

However, I doubt this is the the root-cause here because the the `hardwareModel` stays the same for other VMs affected by `tart set --random-serial`, yet their hardware UUIDs and provisioning UDIDs change.

Here are two VMs with the same `hardwareModel` (!), different `ecid` (observed on host) and different identifiers (observed from within a VM):

<img width="1136" alt="Screenshot 2024-12-03 at 14 53 32" src="https://github.com/user-attachments/assets/8ba6fd97-6bad-478d-83ab-84bfe685cb04">

<img width="1136" alt="Screenshot 2024-12-03 at 14 54 48" src="https://github.com/user-attachments/assets/a397977e-2fbe-4234-808d-48268fc8bf6d">


Above you've said that:

>when enrolling into a MDM server we need a unique UUID and the hardwareModel seems to be the key here.

1. Does this UUID (or UDID) seen on the MDM's side equal to the UUID (or UDID)  you can see in `System Information` inside of a VM?

2. How does the MDM calculate the UUID (or UDID)?

**Reply by motionbug** [reply]
It's the UUID that you can find in system settings or about this mac. 

I will dig further into this and see what I can find! But the only difference I can find was this. 

Out for the day will tackle this tomorrow morning. 

Thank you so much for your help. I wish I was a bit smarter on this subject matter so not to waste your time. 

Again thank you

**Reply by edigaryev** [reply]
>It's the UUID that you can find in system settings or about this mac.

Then `tart set --random-serial` should change it 👌 

If you observe a different behavior, please describe how you've arrived there, and we'll take a look.

**Comment by motionbug** [comment]
@edigaryev Thank you for all your help

I did a bunch of testing. There is no problem with the `tart set --random-serial --random-mac` command, I am getting different UUIDs within my MDM.

I think I found the reason and wonder why it doesn't affect your packer images but there is a keychain issue. Within the system there are:

/Library/Keychains/apsd.keychain
/Users/admin/Library/Keychains/{UUID}/
/Users/admin/Library/Keychains/login.keychain

When I manually delete all these files and do a reboot, I can enrol the device with a MDM profile.

I just now added a shell provisioner to my packer build file to delete the files and then will see if I can enrol the device.

**Comment by motionbug** [comment]
@edigaryev its the keychains for some reason..

Once I blow away all the keychains i can reboot and enroll. With your packer images, do you do anything with the keychains? What hardware builds them?

Thanks

**Reply by edigaryev** [reply]
>With your packer images, do you do anything with the keychains? What hardware builds them?

Which `ghcr.io/cirruslabs/macos-*` image are you using, exactly?

We do some [certificate-related modifications](https://github.com/cirruslabs/macos-image-templates/blob/d3ad77c873284a7424b7b286ffcbc1bf47ed5629/templates/xcode.pkr.hcl#L194-L206) for Xcode images.

We also [install CA certificates](https://github.com/cirruslabs/macos-image-templates/blob/d3ad77c873284a7424b7b286ffcbc1bf47ed5629/templates/base.pkr.hcl#L62) in base images.

All of the Packer templates we use for building these images are available here: https://github.com/cirruslabs/macos-image-templates.

**Reply by motionbug** [reply]
I am using your templates and when I build from your template vanilla or base, just build one now and I have the enrolment issue with the keychains that I do not when pulling from the OCI repo...

SO STRANGE

**Comment by motionbug** [comment]
![jamf 2024-12-11-13 51@2x](https://github.com/user-attachments/assets/2b004a05-bbc9-4208-ae92-8868290f73ac)
![jamf 2024-12-11-13 50@2x](https://github.com/user-attachments/assets/411e182f-e177-4d5f-b90b-3a36133ae7f2)

**Reply by motionbug** [reply]
Used this template https://github.com/cirruslabs/macos-image-templates/blob/master/templates/vanilla-sequoia.pkr.hcl

**Comment by motionbug** [comment]
If I pull this image from the repo, no issues with the enrollment, as in don't have to delete the keychain. If build locally issues

**Reply by edigaryev** [reply]
Could you try with a vanilla image?

Or by just manually creating a macOS VM from IPSW, using `tart create --from-ipsw=latest`?

Would the issue still persist?

**Reply by motionbug** [reply]
Tried base and vanilla

manually creating a VM with an IPSW, its like packer. No enrolment due to keychains

I am going to try something else. I am going to build a packer image on a different computer, upload it my OCI host and then pull it down on another computer, maybe it has to do with some odd certificate signing, since running on a different computer.. the specific certificate is MOOT and it then will jsut create a new one? Going to try
